myia/prim/shape_inferrers.py

```diff
@@ -0,0 +1,161 @@
+"""Definition of shape inference for primitives."""
+
+import operator
+from functools import partial, reduce
+
+from ..infer import ANYTHING, GraphInferrer, register_inferrer, \
+    Track, MyiaShapeError, Inferrer
+from ..ir import Graph
+from ..dtype import Array
+
+from . import ops as P
+from .ops import Primitive
+
+
+def prod(iterable):
+    """Return the product of the elements of the iterator."""
+    return reduce(operator.mul, iterable, 1)
+
+
+shape_inferrer_constructors = {}
+
+
+class ScalarShapeInferrer(Inferrer):
+    """Shape inferrer for all primitves that don't take arrays."""
+
+    def __init__(self, engine):
+        """Initialize the ScalarShapeInferrer."""
+        super().__init__(engine, 'scalar_shape_inferrer')
+
+    async def __call__(self, *args):
+        """Since no arrays are involved, the shape is always ()."""
+        return ()
```

Comment:
I would find it slightly confusing that the shape of a list or tuple would be `()`.
Could we have a way to distinguish arithmetic scalars from other non-arrays?

Reply:
Originally I wanted there to be no shape for all non-arrays, but that confuses the inferrer in numerous spots, so I settled on ().

I don't think that distinguishing scalars from lists and tuple is relevant since we don't support the numpy-style automatic cast to array from those.  I could make it clearer that the shape is only relevant for things that are typed `Array()`